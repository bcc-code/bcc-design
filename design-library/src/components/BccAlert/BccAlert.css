@layer components {
    .bcc-alert {
        @apply text-body-small relative px-4 py-3.5 rounded-md flex items-start gap-x-2.5 gap-y-1;
        /* Default info context */
        @apply bg-info text-info border border-info;
    }

    /* Context */
    .bcc-alert-success {
        @apply bg-success text-success border-success;
    }
    .bcc-alert-warning {
        @apply bg-warning text-warning border-warning;
    }
    .bcc-alert-danger {
        @apply bg-danger text-danger border-danger;
    }
    .bcc-alert-neutral {
        @apply bg-brand-50 text-brand-600 border-brand-300;
    }

    /* Icon */
    .bcc-alert-icon {
        @apply h-5 w-5 shrink-0;
    }

    /* Close button */
    .bcc-alert-close-button {
        @apply absolute right-4 top-4;
    }
    .bcc-alert-close-icon {
        @apply h-4 w-4;
    }

    /* Title */
    .bcc-alert-title {
        @apply text-heading-small;
    }

    /* Content */
    .bcc-alert-content {
        @apply pr-4;
    }
}
