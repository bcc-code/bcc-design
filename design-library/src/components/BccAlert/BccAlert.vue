<script setup lang="ts">
import {
  CloseIcon,
  ErrorFillIcon,
  InfoFillIcon,
  TaskAltIcon,
  WarningFillIcon,
  NotificationsFillIcon,
} from "@bcc-code/icons-vue";
import type { Component } from "vue";

type Props = {
  context?: keyof typeof contexts;
  icon?: boolean | Component;
  title?: string;
  closeButton?: boolean;
  open?: boolean;
};

withDefaults(defineProps<Props>(), {
  context: "info",
  icon: false,
  closeButton: false,
  open: true,
});

const emit = defineEmits(["close"]);

const contexts = {
  info: {
    class: "",
    icon: InfoFillIcon,
  },
  success: {
    class: "bcc-alert-success",
    icon: TaskAltIcon,
  },
  warning: {
    class: "bcc-alert-warning",
    icon: WarningFillIcon,
  },
  danger: {
    class: "bcc-alert-danger",
    icon: ErrorFillIcon,
  },
  neutral: {
    class: "bcc-alert-neutral",
    icon: NotificationsFillIcon,
  },
};
</script>

<template>
  <Transition name="bcc-fade">
    <div v-if="open" class="bcc-alert" :class="contexts[context].class">
      <component
        v-if="icon"
        :is="typeof icon === 'boolean' ? contexts[context].icon : icon"
        class="bcc-alert-icon"
      />

      <div class="bcc-alert-content">
        <h3 v-if="title" class="bcc-alert-title">{{ title }}</h3>
        <p><slot></slot></p>
      </div>
      <button
        @click.prevent="emit('close')"
        v-if="closeButton"
        aria-label="Close alert"
        class="bcc-alert-close-button"
      >
        <CloseIcon class="bcc-alert-close-icon" />
      </button>
    </div>
  </Transition>
</template>
