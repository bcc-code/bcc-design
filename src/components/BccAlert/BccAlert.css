@layer components {
    .bcc-alert {
        @apply text-body-small px-4 py-3.5 rounded-md bg-info text-info border border-info;
    }

    /* Context */
    .bcc-alert-success {
        @apply bg-success text-success border-success;
    }
    .bcc-alert-warning {
        @apply bg-warning text-warning border-warning;
    }
    .bcc-alert-danger {
        @apply bg-danger text-danger border-danger;
    }
}
