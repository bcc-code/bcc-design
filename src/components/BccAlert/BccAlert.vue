<script setup lang="ts">
import { CloseIcon, ErrorIcon, InfoIcon, TaskAltIcon, WarningIcon } from "@bcc-code/icons-vue";
import { TransitionRoot } from "@headlessui/vue";

type Props = {
  context?: keyof typeof contexts;
  icon?: boolean;
  title?: string;
  closeButton?: boolean;
  open?: boolean;
};

withDefaults(defineProps<Props>(), {
  context: "info",
  icon: false,
  closeButton: false,
  open: true,
});

const emit = defineEmits(["close"]);

const contexts = {
  info: {
    class: "",
    icon: InfoIcon,
  },
  success: {
    class: "bcc-alert-success",
    icon: TaskAltIcon,
  },
  warning: {
    class: "bcc-alert-warning",
    icon: WarningIcon,
  },
  danger: {
    class: "bcc-alert-danger",
    icon: ErrorIcon,
  },
};
</script>

<template>
  <TransitionRoot
    :show="open"
    enter="transition-opacity duration-100"
    enter-from="opacity-0"
    enter-to="opacity-100"
    leave="transition-opacity duration-150"
    leave-from="opacity-100"
    leave-to="opacity-0"
  >
    <div class="bcc-alert" :class="contexts[context].class">
      <component v-if="icon" :is="contexts[context].icon" class="bcc-alert-icon" />

      <h3 v-if="title" class="bcc-alert-title">{{ title }}</h3>
      <p class="bcc-alert-content">
        <slot />
      </p>
      <button
        @click.prevent="emit('close')"
        v-if="closeButton"
        aria-label="Close alert"
        class="bcc-alert-close-button"
      >
        <CloseIcon class="bcc-alert-close-icon" />
      </button>
    </div>
  </TransitionRoot>
</template>
