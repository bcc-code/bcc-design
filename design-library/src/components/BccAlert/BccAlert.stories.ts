import BccButton from "../BccButton/BccButton.vue";
import BccAlert from "./BccAlert.vue";

import type { Meta, StoryFn } from "@storybook/vue3";
import { Face2Icon } from "@bcc-code/icons-vue";

export default {
  title: "Components/BccAlert",
  component: BccAlert,
  argTypes: {
    context: {
      options: ["info", "success", "warning", "danger", "neutral"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccAlert>;

const Template: StoryFn<typeof BccAlert> = (args) => ({
  components: { BccAlert, BccButton },
  setup() {
    return { args };
  },
  template: `
    <BccAlert v-bind="args" @close="args.open = false">
      {{ args.slotDefault }}
    </BccAlert>

    <BccButton variant="secondary" @click="args.open = true" v-if="!args.open">
      Show alert
    </BccButton>
  `,
});

export const Example = Template.bind({});
Example.args = {
  open: true,
  context: "success",
  icon: true,
  closeButton: true,
  title: "Well done!",
  slotDefault: "Successfully uploaded",
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccAlert
  title="Well done!"
  icon
  closeButton
  context="success"
  :open="showSuccessAlert"
  @close="showSuccessAlert = false"
>
  Successfully uploaded
</BccAlert>
    `,
    },
  },
};

/**
 * Set the `context` prop to adjust the color of the alert. Pass content to the `default` slot to display text.
 */
export const Basic: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="success"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="warning"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="danger"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="neutral"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
    </div>
  `,
});

/**
 * Set the `icon` prop to render an icon determined by the `context`. Passing a component as a prop `:icon="FaceIcon"` will override the default icon.
 */
export const WithIcon: StoryFn<typeof BccAlert> = (args) => ({
  components: { BccAlert },
  setup() {
    return { args, Face2Icon };
  },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert icon><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="success"><strong>Oh snap!</strong> You might want to check this out! Or maybe just let it be</BccAlert>
      <BccAlert icon context="warning"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="danger"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="neutral"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert :icon="Face2Icon" context="neutral">Look at this custom icon I added!</BccAlert>
    </div>
  `,
});

/**
 * Set the `title` prop to render a title above the context, optionally next to the icon
 */
export const WithTitle: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert title="Well done!" icon>You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="success">You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="warning">You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="danger">You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="neutral">You might want to check this out!</BccAlert>
    </div>
  `,
});

/**
 * Set the `closeButton` prop to render a close button. Clicking this will emit a `close` event, which you can use to change the `open` prop to `false` to hide the alert.
 */
export const WithCloseButton: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert closeButton icon><strong>Aww yeah</strong>, you successfully read this important alert message. This example text is going to run a bit longer so that you can see how spacing within an alert works with this kind of content.</BccAlert>
      <BccAlert closeButton icon>You might want to check this out!</BccAlert>
      <BccAlert closeButton title="Well done!">You might want to check this out!</BccAlert>
      <BccAlert closeButton title="Well done!" icon>Aww yeah, you successfully read this important alert message. This example text is going to run a bit longer so that you can see how spacing within an alert works with this kind of content.</BccAlert>
    </div>
  `,
});
