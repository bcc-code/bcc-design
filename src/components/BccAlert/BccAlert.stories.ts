import BccAlert from "./BccAlert.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Components/BccAlert",
  component: BccAlert,
  argTypes: {
    context: {
      options: ["info", "success", "warning", "danger"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccAlert>;

const Template: StoryFn<typeof BccAlert> = (args) => ({
  components: { BccAlert },
  setup() {
    return { args };
  },
  template: `
    <BccAlert v-bind="args">
      {{ args.slotDefault }}
    </BccAlert>
  `,
});

export const Example = Template.bind({});
Example.args = {
  context: "success",
  icon: true,
  slotDefault: "Successfully uploaded",
  title: "Well done!",
};

/**
 * Set the `context` prop to adjust the color of the alert. Pass content to the `default` slot to display text.
 */
export const Basic: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="success"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="warning"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert context="danger"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
    </div>
  `,
});

/**
 * Set the `icon` prop to render an icon determined by the `context`
 */
export const WithIcon: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert icon><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="success"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="warning"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
      <BccAlert icon context="danger"><strong>Oh snap!</strong> You might want to check this out!</BccAlert>
    </div>
  `,
});

/**
 * Set the `title` prop to render a title above the context, optionally next to the icon
 */
export const WithTitle: StoryFn<typeof BccAlert> = () => ({
  components: { BccAlert },
  template: `
    <div class="flex flex-col space-y-4">
      <BccAlert title="Well done!" icon>You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="success">You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="warning">You might want to check this out!</BccAlert>
      <BccAlert title="Well done!" icon context="danger">You might want to check this out!</BccAlert>
    </div>
  `,
});
