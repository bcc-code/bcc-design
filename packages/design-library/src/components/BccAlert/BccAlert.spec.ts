import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccAlert from "./BccAlert.vue";

describe("BccAlert", () => {
  it("renders an alert", () => {
    const wrapper = mount(BccAlert, {
      props: { title: "Alert title" },
      slots: {
        default: "Alert content",
      },
      global: {
        renderStubDefaultSlot: true,
      },
      shallow: true,
    });

    expect(wrapper.text()).toContain("Alert title");
    expect(wrapper.text()).toContain("Alert content");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("can close", async () => {
    const wrapper = mount(BccAlert, {
      props: { open: true, closeButton: true },
      slots: {
        default: "Alert content",
      },
      global: {
        renderStubDefaultSlot: true,
      },
      shallow: true,
    });

    await wrapper.find(".bcc-alert-close-button").trigger("click");

    expect(wrapper.emitted("close")?.length).toBe(1);
  });
});
