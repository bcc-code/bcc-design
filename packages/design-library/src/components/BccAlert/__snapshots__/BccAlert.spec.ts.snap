// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccAlert > renders an alert 1`] = `
"<anonymous-stub as=\\"div\\" show=\\"true\\" unmount=\\"true\\" appear=\\"false\\" enter=\\"transition-opacity duration-100\\" enterfrom=\\"opacity-0\\" enterto=\\"opacity-100\\" entered=\\"\\" leave=\\"transition-opacity duration-150\\" leavefrom=\\"opacity-100\\" leaveto=\\"opacity-0\\">
  <div class=\\"bcc-alert\\">
    <!--v-if-->
    <h3 class=\\"bcc-alert-title\\">Alert title</h3>
    <p class=\\"bcc-alert-content\\">Alert content</p>
    <!--v-if-->
  </div>
</anonymous-stub>"
`;
